sdk/appcenter-distribute/src/main/java/com/microsoft/appcenter/distribute/download/http/HttpConnectionReleaseDownloader.java

```diff
@@ -0,0 +1,242 @@
+import static com.microsoft.appcenter.distribute.DistributeConstants.LOG_TAG;
+import static com.microsoft.appcenter.distribute.DistributeConstants.PREFERENCE_KEY_DOWNLOADED_RELEASE_FILE;
+
+/**
+ * Downloads new releases directly via HttpsURLConnection for Android versions prior to 5.0.
+ */
+public class HttpConnectionReleaseDownloader implements ReleaseDownloader {
+
+    /**
+     * Context.
+     */
+    private final Context mContext;
+
+    /**
+     * Release to download.
+     */
+    private final ReleaseDetails mReleaseDetails;
+
+    /**
+     * Listener of download status.
+     */
+    private final ReleaseDownloader.Listener mListener;
+
+    public HttpConnectionReleaseDownloader(@NonNull Context context, @NonNull ReleaseDetails releaseDetails, @NonNull ReleaseDownloader.Listener listener) {
+        mContext = context;
+        mReleaseDetails = releaseDetails;
+        mListener = listener;
+    }
+
+    /**
+     * A file that used to write downloaded package.
+     */
+    private File mTargetFile;
+
+    /**
+     * Progress notification builder.
+     * It must be stored to avoid notification flickering.
+     */
+    private Notification.Builder mNotificationBuilder;
+
+    private HttpDownloadFileTask mHttpDownloadFileTask;
+
+    @Nullable
+    private File getTargetFile() {
+        if (mTargetFile == null) {
+            File downloadsDirectory = mContext.getExternalFilesDir(Environment.DIRECTORY_DOWNLOADS);
+            if (downloadsDirectory != null) {
+                mTargetFile = new File(downloadsDirectory, mReleaseDetails.getReleaseHash() + ".apk");
+            }
+        }
+        return mTargetFile;
+    }
+
+    private NotificationManager getNotificationManager() {
+        return (NotificationManager) mContext.getSystemService(Context.NOTIFICATION_SERVICE);
+    }
+
+    /**
+     * Get progress notification builder.
+     *
+     * @see #mNotificationBuilder
+     */
+    @NonNull
+    @SuppressWarnings({"deprecation", "RedundantSuppression"})
+    private Notification.Builder getNotificationBuilder() {
+        if (mNotificationBuilder == null) {
+            mNotificationBuilder = new Notification.Builder(mContext);
+        }
+        return mNotificationBuilder;
+    }
+
+    @Override
+    public synchronized boolean isDownloading() {
+        return mHttpDownloadFileTask != null;
+    }
+
+    @NonNull
+    @Override
+    public ReleaseDetails getReleaseDetails() {
+        return mReleaseDetails;
+    }
+
+    @Override
+    public void resume() {
+        File targetFile = getTargetFile();
+        if (targetFile == null) {
+            mListener.onError("Cannot access to downloads folder. Shared storage is not currently available.");
+            return;
+        }
+
+        /* Check if we have already downloaded the release. */
+        String downloadedReleaseFilePath = SharedPreferencesManager.getString(PREFERENCE_KEY_DOWNLOADED_RELEASE_FILE, null);
```

Comment:
Strict mode violation here. Distribute.resumeDownload is calling this from activity onResume which is UI thread.

Reply:
Did you see it in the report?
We read `SharedPreferences` in a few places on develop. Good example is `completeWorkflow` that **writes** to `SharedPreferences` (it's mush slower).
Once loaded, `SharedPreferences` are singletons and cached process-wide. So, I guess this cache will work here.
Looking at other places with similar issue...

Reply:
Good example of this "violation" is `resumeDistributeWorkflow`

Reply:
This particular issue is fixed